src/REVBasicDeployer.sol

```diff
@@ -500,40 +556,80 @@ contract REVBasicDeployer is ERC165, IREVBasicDeployer, IJBRulesetDataHook, IERC
+}
+
+import {IBPSuckerDeployer} from "@bananapus/suckers/src/interfaces/IBPSuckerDeployer.sol";
+
+/// @custom:member deployer The address to deploy a sucker from for a particular chain pair.
+/// @custom:member tokenConfigurations Information about how the chains will connect.
+struct BPSuckerDeployerConfig {
+    IBPSuckerDeployer deployer;
+    BPTokenConfig[] tokenConfigurations;
+}
+
+interface IBPSuckerRegistry {
+    function suckersOf(uint256 projectId) external view returns (IBPSucker[] memory);
+    function suckerDeployerIsAllowed(address deployer) external view returns (bool);
+
+    function allowSuckerDeployer(address deployer) external;
+    function deploySuckersFor(
+        uint256 projectId,
+        bytes32 salt,
+        REVSuckerDeployerConfig[] memory configurations
+    )
+        external;
+}
+
+contract BPSuckerRegistry is IBPSuckerRegistry {
+    mapping(uint256 projectId => IBPSucker[]) internal _suckersOf;
+
+    mapping(address deployer => bool) public suckerDeployerIsAllowed;
+
+    function suckersOf(uint256 projectId) external view returns (IBPSucker[] memory) {
+        return _suckersOf[projectId];
+    }
+
+    function allowSuckerDeployer(address deployer) public override {
+        // TODO onlyOwner // jbdao.
+        suckerDeployerIsAllowed[deployer] = true;
+    }
+
+    function deploySuckersFor(
+        uint256 projectId,
+        bytes32 salt,
+        REVSuckerDeployerConfig[] memory configurations
+    )
+        public
+        override
+    {
+        // TODO requirePermissions from owner of projectId or Operator of a new DEPLOY_SUCKERS permission.
+
         // Keep a reference to the number of sucker deployers.
-        uint256 numberOfSuckerDeployers = suckerDeploymentConfiguration.deployerConfigurations.length;
+        uint256 numberOfSuckerDeployers = configurations.length;
 
         // Keep a reference to the sucker deploy being iterated on.
-        REVSuckerDeployerConfig memory deployerConfiguration;
+        REVSuckerDeployerConfig memory configuration;
 
         for (uint256 i; i < numberOfSuckerDeployers; i++) {
-            //  Get the configuration being iterated on.
-            deployerConfiguration = suckerDeploymentConfiguration.deployerConfigurations[i];
+            // Get the configuration being iterated on.
+            configuration = configurations[i];
 
-            // Create the sucker.
-            IBPSucker sucker = deployerConfiguration.deployer.createForSender({
-                _localProjectId: revnetId,
-                _salt: suckerDeploymentConfiguration.salt
-            });
+            // Make sure the deployer is allowed.
+            if (!suckerDeployerIsAllowed[address(configuration.deployer)]) revert();
 
-            // Store the sucker.
-            suckersOf[revnetId].push(sucker);
+            // Create the sucker.
+            IBPSucker sucker = configuration.deployer.createForSender({_localProjectId: projectId, _salt: salt});
```

Comment:
MINT_TOKENS is handled by the `hasMintPermissionFor` spec from the data hook. USE_ALLOWANCE needs adding!

Let's see how this shapes up once we get the registry tight.

Reply:
deployer and the sucker both need the permission?

Reply:
Yeah, the deployer needs it because it might have the feeless flag. But the deployer checks that the sucker that is calling it also has the `USE_ALLOWANCE` permission for the project.